### PR TITLE
Features/add graph to table

### PR DESCRIPTION
**Pull request by akinolu52**
## What
*Add Graph tab to the Table Screen*

### CODE REVIEW

File: re_data_ui/src/components/LineageGraph.tsx [hunk 1]
```diff
@@ -1,36 +1,50 @@
...
-  }
-  const getNetwork = (n: Network) => {
+  // console.log('data: ', data);
+  useEffect(() => {
+    if (network && model && data) {
+      network?.body?.data?.nodes?.clear();
+      network?.body?.data?.nodes?.add(data.nodes);
+      network?.body?.data?.edges?.clear();
+      network?.body?.data?.edges?.add(data.edges);
+    }
```
Comment (dejii) [inline re_data_ui/src/components/LineageGraph.tsx:35]: I would prefer you clear switch the order a bit. To clear nodes and edges first before adding the new ones. Looks to me like there's a brief moment where have new nodes and old edges.

File: re_data_ui/src/components/LineageGraph.tsx [hunk 2]
```diff
@@ -45,9 +59,19 @@ const LineageGraph: React.FC<LineageGraphProps> = (
...
+    if (!overviewDataLoaded) return false; // allow re-renders if we've not loaded the graph
+    // allow re-renders if we've not loaded the graph
+    // Return true indicating that the props are always
+    // equal to avoid re-rendering of lineage graph since it's static
+    return true;
+  } if (!showModelDetails) {
+    if (JSON.stringify(prevData) !== JSON.stringify(data)) return false;
+    return true;
+  }
+  return false;
```
Comment (dejii) [inline re_data_ui/src/components/LineageGraph.tsx:76]: do we still need to control the re rendering using this `propsAreEqual` function? Now that we have a function that clears nodes and edges before adding inside the `LineageGraph` component.

File: re_data_ui/src/contexts/redataOverviewContext.ts [hunk 3]
```diff
@@ -187,4 +190,5 @@ export const RedataOverviewContext = React.createContext<OverviewData>({
   generated_at: '',
   tests: [],
   loading: true,
+  dbtMapping: {},
```
Comment (dejii) [inline re_data_ui/src/contexts/redataOverviewContext.ts:193]: not sure what name we can use here. `dbtNodeIdMapping`? @mateuszklimek any suggestions?

File: re_data_ui/src/pages/Dashboard.tsx [hunk 4]
```diff
@@ -1,16 +1,17 @@
+/* eslint-disable comma-dangle */
```
Comment (dejii) [inline re_data_ui/src/pages/Dashboard.tsx:1]: editor is still stubborn? 😅

File: re_data_ui/src/partials/Graph.tsx [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+    const details = allNodes[modelTitle];
+    const modelId = `${details.database}.${details.schema}.${details.name}`.toLowerCase();
+    graph.nodes.push({
+      id: modelId,
+      label: details.name,
+      shape: 'box',
+      color: resourceTypeColors[details.resource_type],
+    });
+
+    if (modelParentNodes.length) {
```
Comment (dejii) [inline re_data_ui/src/partials/Graph.tsx:90]: is there a need for this check? `.forEach()` won't iterate if the array is empty.

File: re_data_ui/src/partials/Graph.tsx [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        setGraph(generateGraph(overview));
+      }
+    }
+  };
+
+  useEffect(() => {
+    (async function anyNameFunction() {
+      await updateGraph();
+    }());
+  }, [modelName, overview]);
```
Comment (dejii) [inline re_data_ui/src/partials/Graph.tsx:209]: what's the reason behind using an `IIFE` here?

File: re_data_ui/src/partials/Graph.tsx [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+            arrows: 'to',
+          };
+          graph.edges.push(edge);
+        }
+      });
+    }
+  } else {
+    Object.entries(allNodes).forEach(([, details]) => {
+      if (supportedResTypes.has(details.resource_type) && details.package_name !== 're_data') {
+        const modelId = `${details.database}.${details.schema}.${details.name}`.toLowerCase();
```
Comment (dejii) [inline re_data_ui/src/partials/Graph.tsx:143]: we can refactor this logic into a function since we're using it in more than one place. what do you think?

Comment (dejii) [reply re_data_ui/src/partials/Graph.tsx:143]: i.e generating the identifier

File: re_data_ui/src/components/LineageGraph.tsx (hunk 1 above)
Comment (akinolu52) [reply re_data_ui/src/components/LineageGraph.tsx:35]: okay

File: re_data_ui/src/components/LineageGraph.tsx (hunk 2 above)
Comment (akinolu52) [reply re_data_ui/src/components/LineageGraph.tsx:76]: i used that as a way to double check but on second thought, there's actually no need

File: re_data_ui/src/partials/Graph.tsx (hunk 5 above)
Comment (akinolu52) [reply re_data_ui/src/partials/Graph.tsx:90]: yeah, thanks

File: re_data_ui/src/partials/Graph.tsx (hunk 7 above)
Comment (akinolu52) [reply re_data_ui/src/partials/Graph.tsx:143]: okay, i'll convert it into an helper function

File: re_data_ui/src/components/LineageGraph.tsx (hunk 2 above)
Comment (akinolu52) [reply re_data_ui/src/components/LineageGraph.tsx:76]: @dejii removing this causes re-render once a node has been selected

File: re_data_ui/src/contexts/redataOverviewContext.ts (hunk 3 above)
Comment (mateuszklimek) [reply re_data_ui/src/contexts/redataOverviewContext.ts]: Hm... I think dbtMapping is good for me :)

File: re_data_ui/src/components/LineageGraph.tsx (hunk 2 above)
Comment (dejii) [reply re_data_ui/src/components/LineageGraph.tsx:75]: that's strange.

Comment (akinolu52) [reply re_data_ui/src/components/LineageGraph.tsx:75]: i re-checked

removing this prevent the data props from being updated. @dejii

File: re_data_ui/src/contexts/redataOverviewContext.ts (hunk 3 above)
Comment (akinolu52) [reply re_data_ui/src/contexts/redataOverviewContext.ts]: i'll change it back to dbtMapping

File: re_data_ui/src/partials/Graph.tsx (hunk 6 above)
Comment (akinolu52) [reply re_data_ui/src/partials/Graph.tsx]: removed, was testing a different approach...

File: re_data_ui/src/components/LineageGraph.tsx (hunk 2 above)
Comment (dejii) [reply re_data_ui/src/components/LineageGraph.tsx:75]: 👍

**Pull request merged**